### PR TITLE
FIx Redundant Cast incorrect for SimpleArguments #44516

### PR DESCRIPTION
**Pull request by paul1956**
Add tests for Casts in SimpleArguments

Fixes issues #44516

### CODE REVIEW

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Utilities/CastAnalyzer.vb [hunk 1]
```diff
@@ -68,7 +68,6 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Extensions
                     Return semanticModel.GetTypeInfo(parentExpression, cancellationToken).Type
                 End If
             End If
-
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Utilities/CastAnalyzer.vb:71]: undo

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Utilities/CastAnalyzer.vb [hunk 2]
```diff
@@ -113,6 +112,14 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Extensions
                 Return semanticModel.GetTypeInfo(otherExpression).Type
             End If
 
+            Dim parentSimpleArgument = TryCast(parent, SimpleArgumentSyntax)
+            If parentSimpleArgument IsNot Nothing Then
+                If TypeOf parentSimpleArgument.Expression Is CastExpressionSyntax OrElse
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Utilities/CastAnalyzer.vb:117]:
```suggestion
            If parentSimpleArgument?.Expression Is CastExpressionSyntax OrElse
```

**Comment by CyrusNajmabadi** [comment]
This is failing the test: TestVisualBasic_DoNotRemove_NecessaryCastForParamsArgument

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Utilities/CastAnalyzer.vb (hunk 1 above)
Comment (paul1956) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Utilities/CastAnalyzer.vb]: done

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Utilities/CastAnalyzer.vb (hunk 2 above)
Comment (paul1956) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Utilities/CastAnalyzer.vb]: done

**Comment by paul1956** [comment]
@CyrusNajmabadi I believe TestVisualBasic_DoNotRemove_NecessaryCastForParamsArgument should fail as written. If I copy the result into Visual Studio and remove the DirectCast Visual Studio is happy with the results. If you agree I will correct or remove the test.

**Comment by CyrusNajmabadi** [comment]
The test appears to be correct. If you remove the cast, the code will still compile, but the semantics will be different, leading to different runtime behavior.

**Comment by CyrusNajmabadi** [comment]
Please update the pr to handle the new car, without regressing the existing cases. Thanks!

File: src/Analyzers/VisualBasic/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.vb [hunk 3]
```diff
@@ -31,14 +31,46 @@ End Module
...
+
+        <Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnnecessaryCast)>
+        Public Async Function TestDontRemoveCastSimpleArgument2() As Task
+            Dim markup =
+<File>
+Option Strict On
+Imports System.Drawing
+Module M
+    Sub Main()
+       Dim y As New Point([|CInt(1.1), 1)|]
```
Comment (CyrusNajmabadi) [inline src/Analyzers/VisualBasic/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.vb:58]: the selection span is a bit strange here.  was it intentional?

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Utilities/CastAnalyzer.vb [hunk 4]
```diff
@@ -132,6 +138,11 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Extensions
             If speculatedExpressionOuterType Is Nothing Then
                 Return Nothing
             End If
+            Stop
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Utilities/CastAnalyzer.vb:141]: we cannot check in 'stop' statements.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Utilities/CastAnalyzer.vb [hunk 5]
```diff
@@ -132,6 +138,11 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Extensions
...
+            If outerSpeculatedExpression.IsParentKind(SyntaxKind.SimpleArgument) Then
+                speculatedExpressionOuterType = outerType
+                Return _semanticModel.ClassifyConversion(_castExpressionNode.WalkDownParentheses(), speculatedExpressionOuterType)
+            End If
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Utilities/CastAnalyzer.vb:145]: 1. can we doc why we're doing this?
2. are there tests that validate hte behavior when the WalkDownParentheses is triggered?

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Utilities/CastAnalyzer.vb:145]: why does the code above use `speculationAnalyzer.SpeculativeSemanticModel` but this code directly accesses _semanticModel?

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Utilities/CastAnalyzer.vb [hunk 6]
```diff
@@ -266,13 +277,13 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Extensions
...
-                ' to the outer type, we need to make sure that the same user-defined conversion will be 
+                ' to the outer type, we need to make sure that the same user-defined conversion will be
                 ' called if the cast is removed.
                 If castToOuterType.IsUserDefined OrElse expressionToCastType.IsUserDefined Then
                     Return (HaveSameUserDefinedConversion(expressionToCastType, expressionToOuterType) OrElse
                             HaveSameUserDefinedConversion(castToOuterType, expressionToOuterType)) AndAlso
-                           (UserDefinedConversionIsAllowed(_castNode) AndAlso
-                            Not expressionToCastType.IsNarrowing)
+                           UserDefinedConversionIsAllowed(_castNode) AndAlso
+                            Not expressionToCastType.IsNarrowing
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Utilities/CastAnalyzer.vb:286]: this feels more confusing afterwards.

File: src/Analyzers/VisualBasic/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.vb (hunk 3 above)
Comment (paul1956) [reply src/Analyzers/VisualBasic/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.vb:58]: No it's just lack of understand how to select a span. I will fix.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Utilities/CastAnalyzer.vb (hunk 4 above)
Comment (paul1956) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Utilities/CastAnalyzer.vb:141]: That was for debugging, I will remove.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Utilities/CastAnalyzer.vb (hunk 5 above)
Comment (paul1956) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Utilities/CastAnalyzer.vb:145]: The case that currently fails is a cast inside a SimpleArgument. The function GetSpeculatedExpressionToOuterTypeConversion gets called for SimpleArgments it seems it was written to handle a cast of a ParamArray which according to the tests that fail when I remove it should not have the cast removed. I copied the code (lines 142-145) from the If side of the caller that handles other cases.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Utilities/CastAnalyzer.vb (hunk 6 above)
Comment (paul1956) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Utilities/CastAnalyzer.vb:286]: Just following Visual Studio suggestion. I can put it back.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Utilities/CastAnalyzer.vb (hunk 5 above)
Comment (paul1956) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Utilities/CastAnalyzer.vb:144]: I don't know why speculationAnalyzer.SpeculativeSemanticModel was originally used

Comment (paul1956) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Utilities/CastAnalyzer.vb:144]: WalkDownParentheses is called a lot and there any many tests that hit the inner while loop. If you are asking about  line 143 my 2 new tests don't hit the inner loop.

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Utilities/CastAnalyzer.vb:146]: can you add the tests, and use the sepculative model to match teh rest of the code around you?

Comment (paul1956) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Utilities/CastAnalyzer.vb:146]: The node I need to look at at this point is not in the speculationAnalyzer.SpeculativeSemanticModel and it throws an exception. Updated test to test the while look in WalkDownParentheses.

Comment (paul1956) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Utilities/CastAnalyzer.vb:146]: @CyrusNajmabadi added tests to verify line 143.

Comment (paul1956) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Utilities/CastAnalyzer.vb]: @CyrusNajmabadi  Removed _semanticModel from line 142 section. removed most code I added to function. Special case in now in caller what already used SemanticModel.

**Comment by paul1956** [comment]
All tests pass without any changes and new tests added to test issue and they pass..

File: src/Analyzers/VisualBasic/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.vb [hunk 7]
```diff
@@ -31,14 +31,46 @@ End Module
...
+Option Strict On
+Imports System.Drawing
+Module M
+    Sub Main()
+       Dim y As New Point([|CInt(1.1)|], 1)
+    End Sub
+End Module
+</File>
+
+            Await TestMissingAsync(markup)
```
Comment (CyrusNajmabadi) [inline src/Analyzers/VisualBasic/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.vb:63]: need a test that hits the CastExpressionSyntax case you added as well.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Utilities/CastAnalyzer.vb [hunk 8]
```diff
@@ -113,10 +113,16 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Extensions
                 Return semanticModel.GetTypeInfo(otherExpression).Type
             End If
 
+            Dim parentSimpleArgument = TryCast(parent, SimpleArgumentSyntax)
+            If TypeOf parentSimpleArgument?.Expression Is CastExpressionSyntax OrElse
+               TypeOf parentSimpleArgument?.Expression Is PredefinedCastExpressionSyntax Then
+                Return semanticModel.GetTypeInfo(parentSimpleArgument.Expression, cancellationToken).Type
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Utilities/CastAnalyzer.vb:119]: i see tests for the PredefinedCastExpressionSyntax  case.  but not for CastExpressionSyntax

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Utilities/CastAnalyzer.vb [hunk 9]
```diff
@@ -256,7 +262,13 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Extensions
                     speculatedExpressionOuterType = outerType
                     expressionToOuterType = _semanticModel.ClassifyConversion(_castExpressionNode.WalkDownParentheses(), speculatedExpressionOuterType)
                 Else
-                    expressionToOuterType = GetSpeculatedExpressionToOuterTypeConversion(speculationAnalyzer.ReplacedExpression, speculationAnalyzer, _cancellationToken, speculatedExpressionOuterType)
+                    expressionToOuterType = GetSpeculatedExpressionToOuterTypeConversion(speculationAnalyzer, speculationAnalyzer.ReplacedExpression, outerSpeculatedExpression, _cancellationToken, speculatedExpressionOuterType)
+                    If expressionToOuterType = Nothing AndAlso outerSpeculatedExpression.IsParentKind(SyntaxKind.SimpleArgument) Then
+                        ' If we are here we might be inside a SimpleArgument but it is
+                        ' not part of a ParamArray which is handled above.
+                        speculatedExpressionOuterType = outerType
+                        expressionToOuterType = _semanticModel.ClassifyConversion(_castExpressionNode.WalkDownParentheses(), speculatedExpressionOuterType)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Utilities/CastAnalyzer.vb:270]: i don't see tests that validate the WalkDownParentheses case.

Comment (paul1956) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Utilities/CastAnalyzer.vb:270]: TestDontRemoveCastSimpleArgument1 the parentheses here (System.Math.Floor(1.1)) are walked down

**Comment by 333fred** [comment]
This PR has a _lot_ of unrelated files in it. Is there anything that the compiler team actually needs to review in it?

**Comment by jaredpar** [comment]
> This PR has a lot of unrelated files in it. Is there anything that the compiler team actually needs to review in it?

Agree. This really needs to be cleaned up before we can consider merging. This change shouldn't be impacting 500+ files. In addithion to making our history hard to follow it's not reasonable to review such a change.

**Comment by paul1956** [comment]
@jaredpar @333fred I don't know what happened I changed only 2 files. I could delete my fork/PR and start again with just replacing the 2 changed files.

**Comment by CyrusNajmabadi** [comment]
Did you merge something into your branch?

**Comment by jnm2** [comment]
The last commit has a stray second parent. Looks like an accidental merge to an old version of the same commits.

image

The fix:

1. Undo the last commit while leaving its changes staged.
   ```
   git reset head~ --soft
   ```

2. Create the commit as intended (but this time not as a merge commit).
   ```
   git commit -m 'Add comments explaining which tests hit which issue'
   ```

3. Fix the pull request by force-pushing
   ```
   git push -f
   ```

**Comment by paul1956** [comment]
> Did you merge something into your branch?

yes but I don't know how it happened. It should be fixed now.

**Comment by paul1956** [comment]
Any idea what the correct process is to move https://github.com/dotnet/roslyn/pull/45102 along? It is waiting for 1 final review.

**Comment by paul1956** [comment]
Ping @jaredpar @333fred is there anything still holding this up?

**Comment by CyrusNajmabadi** [comment]
Thanks!
